File: Podcaster/TMPodcast.m
//
//  TMPodcast.m
//  Podcaster
//
//  Created by Tyler Mikev on 3/7/15.
//  Copyright (c) 2015 Tyler Mikev. All rights reserved.
//

#import "TMPodcast.h"
#import "TMPodcastEpisode.h"
#import "TMiTunesResponse.h"

@implementation TMPodcast

@synthesize podcastDescription;
@synthesize feedURLString;
@synthesize podcastImage;
@synthesize episodes;
@synthesize title;

+ (instancetype)initWithDictionary:(NSDictionary *)dictionary {
    TMPodcast *podcast = [TMPodcast new];
    [podcast populateFromXMLDictionary:dictionary];
    
    return podcast;
}

+ (instancetype) initWithiTunesResponse:(TMiTunesResponse *)iTunesResponse {
    
    TMPodcast *podcast = [TMPodcast new];
    
    podcast.title = iTunesResponse.collectionName;
    podcast.feedURLString = [iTunesResponse.feedUrl absoluteString];
    podcast.author = iTunesResponse.artistName;
    podcast.imageURL = iTunesResponse.artworkUrl100;

    return podcast;
}

- (void)populateFromXMLDictionary:(NSDictionary *)xmlDictionary {
    
    NSDictionary *podcastDictionary = xmlDictionary[@"rss"][@"channel"];
    
    self.title = podcastDictionary[@"title"][@"text"];
    self.podcastDescription = podcastDictionary[@"description"][@"text"];
    self.language = podcastDictionary[@"language"][@"text"];
    self.copyright = podcastDictionary[@"copyright"][@"text"];
    self.author = podcastDictionary[@"itunes:author"][@"text"];
    self.subtitle = podcastDictionary[@"itunes:subtitle"][@"text"];
    self.imageURL = [NSURL URLWithString:podcastDictionary[@"itunes:image"][@"href"]];
    
    self.episodes = [self episodesFromXMLDictionary:podcastDictionary];
}


- (NSSet *)episodesFromXMLDictionary:(NSDictionary *)xmlDictionary {
    id itemsElement = xmlDictionary[@"item"];
    NSSet *episodesSet = nil;
    
    if (itemsElement) {
        NSArray *itemsArray = nil;
        if (itemsElement) {
            if ([itemsElement isKindOfClass:[NSDictionary class]]) {
                itemsArray = @[itemsElement];
            } else if ([itemsElement isKindOfClass:[NSArray class]]) {
                itemsArray = itemsElement;
            }
        }
        
        episodesSet = [TMPodcastEpisode episodesFromDictionariesArray:itemsArray forPodcast:self];
    }
   
    return episodesSet;
}



@end

File: Podcaster/TMPodcast.swift
//
//  TMPodcast.swift
//  Podcaster
//
//  Created by Tyler Mikev on 3/22/15.
//  Copyright (c) 2015 Tyler Mikev. All rights reserved.
//

import Foundation

class TMPodcast {
    
    var title : String?
    var subtitle : String?
    var linkURLString : String?
    var podcastDescription : String?
    var language : String?
    var copyright: String?
    var author : String?
    var category : String?
    var imageURLString : String?
    var episodes : [TMPodcastEpisode]?
    
    init (dictionary : NSDictionary) {
        let podcastDictionary = JSON(dictionary)
        self.title = podcastDictionary["title"]["text"].string
        self.subtitle = podcastDictionary["itunes:subtitle"]["text"].string
        self.linkURLString = podcastDictionary["link"]["text"].string
        self.podcastDescription = podcastDictionary["description"]["text"].string
        self.language = podcastDictionary["language"]["text"].string
        self.copyright = podcastDictionary["copyright"]["text"].string
        self.author = podcastDictionary["itunes:author"]["text"].string
        self.category = podcastDictionary["itunes:category"]["text"].string
        self.imageURLString = podcastDictionary["itunes:image"]["href"].string
        self.episodes = self.episodesFromPodcastDictionary(podcastDictionary)
    }
    
    func episodesFromPodcastDictionary(podcastDictionary : JSON) -> [TMPodcastEpisode]? {
        var episodesElement = podcastDictionary["item"]
        var episodeDictionariesArray : [JSON]?
        
        //we need this to be in array form, so make sure this is an array of dictionaries
        if episodesElement.type == .Dictionary {
            episodeDictionariesArray = [episodesElement]
        } else if episodesElement.type == .Array {
            episodeDictionariesArray = episodesElement.array
        }
        
        var episodesArray : [TMPodcastEpisode]?
        if let episodeDictionariesArrayValue = episodeDictionariesArray? {
            for episodeDictionary in episodeDictionariesArrayValue {
                var episode = TMPodcastEpisode(episodeDictionary: episodeDictionary)
                
                if episodesArray == nil {
                    episodesArray = []
                }
                
                episodesArray?.append(episode)
            }
        }
        
        return episodesArray
    }

}